python/tests/test_tables.py

```diff
@@ -2361,6 +2362,134 @@ def test_samples_interface(self):
+            tables.individuals.add_row(
+                flags=individual.flags,
+                location=individual.location,
+                parents=individual.parents,
+                metadata={
+                    "original_id": i,
+                    "original_parents": [int(p) for p in individual.parents],
+                },
+            )
+        tables.sort()
+        return tables
+
+    def test_individual_parent_mapping(self, wf_sim_with_individual_metadata):
+        tables = wf_sim_with_individual_metadata.copy()
+        tables.simplify()
+        ts = tables.tree_sequence()
+        for individual in tables.individuals:
+            for parent, original_parent in zip(
+                individual.parents, individual.metadata["original_parents"]
+            ):
+                if parent != tskit.NULL:
+                    assert (
+                        ts.individual(parent).metadata["original_id"] == original_parent
+                    )
+        assert set(tables.individuals.parents) != {-1}
+
+    def test_shuffled_individual_parent_mapping(self, wf_sim_with_individual_metadata):
+        tables = wf_sim_with_individual_metadata.copy()
+        individuals = tables.individuals
+
+        new_order = np.arange(individuals.num_rows)
+        np.random.default_rng(42).shuffle(new_order)
+
+        def shuffle_ragged_array(array, offset, new_order):
+            new = np.zeros_like(array)
+            new_offset = np.zeros_like(offset)
+            pos = 0
+            for ni, i in enumerate(new_order):
+                start = offset[i]
+                stop = offset[i + 1]
+                new[pos : pos + (stop - start)] = array[start:stop]
+                new_offset[ni] = pos
+                pos += stop - start
+            new_offset[-1] = pos
+            return new, new_offset
+
+        tables.individuals.flags = individuals.flags[new_order]
+        individuals.location, individuals.location_offset = shuffle_ragged_array(
+            individuals.location, individuals.location_offset, new_order
+        )
+        individuals.parents, individuals.parents_offset = shuffle_ragged_array(
+            individuals.parents, individuals.parents_offset, new_order
+        )
+        individuals.metadata, individuals.metadata_offset = shuffle_ragged_array(
+            individuals.metadata, individuals.metadata_offset, new_order
+        )
+
+        # Remap parents
+        new_index = {i: ni for ni, i in enumerate(new_order)}
+        individuals.parents = [
+            -1 if i == -1 else new_index[i] for i in individuals.parents
+        ]
+
+        # Check we have a mixed up order
+        with pytest.raises(
+            tskit.LibraryError,
+            match="Individuals must be provided in an order where"
+            " children are after their parent individuals",
+        ):
+            tables.tree_sequence()
+
+        tables.simplify()
+        metadata = [
```

Comment:
Might be easier make a tree sequence out of it, and let machinery do the metadata processing for you?

Reply:
We can't as the individuals being shuffled violates the requirements for a tree sequence. Hence the `raises` check above.

Reply:
Doesn't ``simplify`` sort this out, or is that a follow on?

Reply:
No, simplify doesn't sort the individuals. You'd have to do a `sort_tables` after, but then you're not just testing simplify.

Reply:
Fair enough :+1: